### PR TITLE
Fix *Plot.style() methods

### LINKED ISSUE #1633: DotPlot `.style()` resets previous value assignments
Multiple calls to `sc.pl.Dotplot.style()` cause resetting of parameters. This issue can be observed as a change in the default colormap shown by #1632  Other methods that set default parameters are also affected like `.add_totals()`.

The following example should show the dots using the `Reds` colormap, but instead it uses the `winter` colormap because the second call sets the color map to `winter` if not given. This double call happens because when `sc.pl.dotplot()` is used (instead of `sc.pl.DotPlot`), internally a call to `.style()` is made and a subsequent explicit calls to `.style()` is required to tune the parameters as suggested in the documentation.

```python
adata = sc.datasets.pbmc68k_reduced()
markers = ['C1QA', 'PSAP', 'CD79A', 'CD79B', 'CST3', 'LYZ']
sc.pl.DotPlot(adata, markers, groupby='bulk_labels').style(cmap='Reds').style(dot_edge_color='black').show()
```
image

The problem is caused by the current implementation of `sc.pl.Dotplot.style()` that set the default parameters as:

```
  def style(
        self,
        cmap: str = DEFAULT_COLORMAP,
        color_on: Optional[Literal['dot', 'square']] = DEFAULT_COLOR_ON,
        dot_max: Optional[float] = DEFAULT_DOT_MAX,
        dot_min: Optional[float] = DEFAULT_DOT_MIN,
        .....
```

Where DEFAULT_* are the default values defined at the beginning of the file (see https://github.com/theislab/scanpy/blob/master/scanpy/plotting/_dotplot.py#L84) 

What is nice about this is that the documentation clearly shows the default values.

The downside is that optional values are assigned a default value that rewrites previous calls to style. Ideally, all optional values should be `None`, then is easy to know if a new value is passed or a previous call has already set a value. But, doing so will remove the defaults from the documentation.

@flying-sheep suggested to use a code he wrote to add default annotations to the documentation

https://github.com/theislab/scanpydoc/blob/875b441212830678cf9fc81c52f5af29bbb8715f/scanpydoc/elegant_typehints/formatting.py#L101-L107

### PR DESCRIPTION
**Pull request by flying-sheep**
- [x] Closes #1633
- [x] Tests included or not required because:

- [ ] Release notes not necessary because:

This adds relies on `_empty` to check which parameters have been specified, and only changes those

### CODE REVIEW

File: src/scanpy/plotting/_dotplot.py [hunk 1]
```diff
@@ -303,18 +301,18 @@ def __init__(
...
-        dot_min: float | None = DEFAULT_DOT_MIN,
-        smallest_dot: float | None = DEFAULT_SMALLEST_DOT,
-        largest_dot: float | None = DEFAULT_LARGEST_DOT,
-        dot_edge_color: ColorLike | None = DEFAULT_DOT_EDGECOLOR,
-        dot_edge_lw: float | None = DEFAULT_DOT_EDGELW,
-        size_exponent: float | None = DEFAULT_SIZE_EXPONENT,
-        grid: float | None = False,
-        x_padding: float | None = DEFAULT_PLOT_X_PADDING,
-        y_padding: float | None = DEFAULT_PLOT_Y_PADDING,
+        cmap: str | None | Empty = _empty,
```
Comment (ilan-gold) [inline src/scanpy/plotting/_dotplot.py]: Why can this be `None` now?

Comment (ilan-gold) [reply src/scanpy/plotting/_dotplot.py]: and others

**Review by ilan-gold — COMMENTED** [review]
Can we add a test for this?

File: src/scanpy/plotting/_dotplot.py (hunk 1 above)
Comment (flying-sheep) [reply src/scanpy/plotting/_dotplot.py]: `self.cmap` was always allowed to be None, the type annotations for `style` were just wrong.

**Comment by flying-sheep** [comment]
@ilan-gold I cleaned up all the style parameters around where `None` is valid or not

**Review by ilan-gold — APPROVED** [review]

File: src/scanpy/plotting/_stacked_violin.py [hunk 2]
```diff
@@ -814,6 +815,7 @@ def stacked_violin(
         var_group_labels=var_group_labels,
         var_group_rotation=var_group_rotation,
         layer=layer,
+        categories_order=order,
```
Comment (ilan-gold) [inline src/scanpy/plotting/_stacked_violin.py]: What prompted this?

File: src/scanpy/plotting/_stacked_violin.py [hunk 3]
```diff
@@ -823,7 +825,9 @@ def stacked_violin(
     )
 
     if dendrogram:
-        vp.add_dendrogram(dendrogram_key=dendrogram)
+        vp.add_dendrogram(
+            dendrogram_key=dendrogram if isinstance(dendrogram, str) else None
```
Comment (ilan-gold) [inline src/scanpy/plotting/_stacked_violin.py]: The typing then of `add_dendrogram` with `dendrogram_key: str | None = None,` seems off, no? 

https://github.com/scverse/scanpy/blob/578e57297e8cfa48e5492e58b45567217a6ae43b/src/scanpy/plotting/_baseplot_class.py#L239C9-L245

**Review by ilan-gold — CHANGES_REQUESTED** [review]

File: src/scanpy/plotting/_stacked_violin.py (hunk 3 above)
Comment (ilan-gold) [reply src/scanpy/plotting/_stacked_violin.py]: Oh apologies, I misread this.

Comment (flying-sheep) [reply src/scanpy/plotting/_stacked_violin.py]: I think I got to the bottom of this. commit incoming

File: src/scanpy/plotting/_stacked_violin.py (hunk 2 above)
Comment (flying-sheep) [reply src/scanpy/plotting/_stacked_violin.py]: just a bug I found, I saw that `order` was unused.

Comment (ilan-gold) [reply src/scanpy/plotting/_stacked_violin.py]: Ok, do you think it needs a test? Some way to check this? Perhaps overkill, though....up to you

**Comment by flying-sheep** [comment]
OK, I undid the `dendrogram` change and instead added the effective typing for all the dendrogram-related functions.

I was confused by the unclear naming convention and thought that `dendrogram` was bool|str|None whereas `dendrogram_key` was str|None, but the names are just used interchangeably.

File: src/scanpy/plotting/_stacked_violin.py (hunk 2 above)
Comment (flying-sheep) [reply src/scanpy/plotting/_stacked_violin.py]: I changed my mind. `categories_order` is the standard name everywhere else so I just removed the unused `order` instead.

File: src/scanpy/plotting/_stacked_violin.py [hunk 4]
```diff
@@ -727,10 +727,6 @@ def stacked_violin(
         See :func:`~seaborn.stripplot`.
     size
         Size of the jitter points.
-    order
-        Order in which to show the categories. Note: if `dendrogram=True`
-        the categories order will be given by the dendrogram and `order`
-        will be ignored.
```
Comment (ilan-gold) [inline src/scanpy/plotting/_stacked_violin.py]: Also,isn't this a breaking change?  This is a public API, no?

File: src/scanpy/plotting/_stacked_violin.py [hunk 5]
```diff
@@ -696,8 +696,10 @@ def stacked_violin(
     jitter: float | bool = StackedViolin.DEFAULT_JITTER,
     size: int | float = StackedViolin.DEFAULT_JITTER_SIZE,
     row_palette: str | None = StackedViolin.DEFAULT_ROW_PALETTE,
-    scale: Literal["area", "count", "width"] = StackedViolin.DEFAULT_DENSITY_NORM,
+    density_norm: Literal["area", "count", "width"] | Empty = _empty,
     yticklabels: bool = StackedViolin.DEFAULT_PLOT_YTICKLABELS,
+    # deprecated
+    scale: Literal["area", "count", "width"] | Empty = _empty,
```
Comment (ilan-gold) [inline src/scanpy/plotting/_stacked_violin.py]: This is from an old PR? #2844? And you noticed it here? Should it be tested?

File: src/scanpy/plotting/_stacked_violin.py (hunk 4 above)
Comment (ilan-gold) [reply src/scanpy/plotting/_stacked_violin.py]: https://scanpy.readthedocs.io/en/stable/generated/scanpy.pl.stacked_violin.html So I see that we have the `categories_order` docstring (not sure how - common args?) but `order` still seems to be a public API?  And now it's gone: https://icb-scanpy--3206.com.readthedocs.build/en/3206/generated/scanpy.pl.stacked_violin.html?

**Review by ilan-gold — CHANGES_REQUESTED** [review]
Might be off on the public API thing, but seems like it would require a minor version bump then....not sure if it's actually relevant anymore to the goal of this PR so might be worth separating out into its own PR.

File: src/scanpy/plotting/_stacked_violin.py (hunk 4 above)
Comment (flying-sheep) [reply src/scanpy/plotting/_stacked_violin.py]: It’s a public API … that used to be completely ignored. I can add a warning like

> `order` is ignored. use `categories_order` instead

File: src/scanpy/plotting/_stacked_violin.py (hunk 5 above)
Comment (flying-sheep) [reply src/scanpy/plotting/_stacked_violin.py]: I’ll start pulling unrelated things like this out!

File: src/scanpy/plotting/_stacked_violin.py (hunk 4 above)
Comment (ilan-gold) [reply src/scanpy/plotting/_stacked_violin.py]: Interesting ok! In that case, yea a warning would be better than removal

Comment (flying-sheep) [reply src/scanpy/plotting/_stacked_violin.py]: OK, moved to #3252

**Review by ilan-gold — APPROVED** [review]

**Pull request merged**